### PR TITLE
Add dynamic templates to bulk indexer item

### PR DESCRIPTION
**Pull request by carsonip**
Make bulk request accept dynamic templates by adding DynamicTemplates field to bulk indexer item. This is only added to bulk indexer, not appender.

### CODE REVIEW

File: bulk_indexer.go [hunk 1]
```diff
@@ -287,18 +288,39 @@ func (b *BulkIndexer) Add(item BulkIndexerItem) error {
...
 	if index != "" {
-		if documentID != "" {
+		if !first {
 			b.jsonw.RawByte(',')
 		}
 		b.jsonw.RawString(`"_index":`)
 		b.jsonw.String(index)
+		first = false
+	}
+	if dynamicTemplates != nil && len(dynamicTemplates) > 0 {
```
Comment (axw) [inline bulk_indexer.go:307]:
```suggestion
	if len(dynamicTemplates) > 0 {
```

**Review by axw — COMMENTED** [review]
Looks good. Could we add a small unit test that validates the _bulk request format?

File: bulk_indexer.go [hunk 2]
```diff
@@ -269,14 +269,15 @@ func (b *BulkIndexer) BytesUncompressedFlushed() int {
 }
 
 type BulkIndexerItem struct {
-	Index      string
-	DocumentID string
-	Body       io.WriterTo
+	Index            string
+	DocumentID       string
+	Body             io.WriterTo
+	DynamicTemplates map[string]string
```
Comment (carsonip) [inline bulk_indexer.go:275]: @axw I was initially a little hesitant to pass in a map here, as Body is a io.WriterTo containing serialized bytes which makes it look inconsistent. But now I'm fine with the idea, as DynamicTemplates does not really require specific data structures and is only a `map[string]string`. WDYT?

Comment (axw) [reply bulk_indexer.go:275]: I think of it like this:
 - `DynamicTemplates` has a known structure, which fits naturally to `map[string]string`
 - `Body` just needs to be JSON, but we shouldn't be too prescriptive about how the JSON encoding is done

So I think it's perfectly fine to use `map[string]string` here.

**Review by axw — APPROVED** [review]
LGTM!

**Pull request merged**